framework/include/utils/SymmetricRankFourTensorImplementation.h

```diff
@@ -0,0 +1,558 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "SymmetricRankFourTensor.h"
+
+// MOOSE includes
+#include "SymmetricRankTwoTensor.h"
+#include "RankFourTensor.h"
+#include "MooseEnum.h"
+#include "MooseException.h"
+#include "MooseUtils.h"
+#include "MatrixTools.h"
+#include "PermutationTensor.h"
+
+#include "libmesh/utility.h"
+#include "libmesh/tensor_value.h"
+#include "libmesh/vector_value.h"
+
+// C++ includes
+#include <iomanip>
+#include <ostream>
+
+namespace MathUtils
+{
+template <>
+void mooseSetToZero<SymmetricRankFourTensorTempl<Real>>(SymmetricRankFourTensorTempl<Real> & v);
+template <>
+void
+mooseSetToZero<SymmetricRankFourTensorTempl<DualReal>>(SymmetricRankFourTensorTempl<DualReal> & v);
+}
+
+template <typename T>
+MooseEnum
+SymmetricRankFourTensorTempl<T>::fillMethodEnum()
+{
+  return MooseEnum("symmetric9 symmetric21  symmetric_isotropic symmetric_isotropic_E_nu  "
+                   "axisymmetric_rz principal orthotropic");
+}
+
+template <typename T>
+SymmetricRankFourTensorTempl<T>::SymmetricRankFourTensorTempl()
+{
+  mooseAssert(Ndim == 3,
+              "SymmetricRankFourTensorTempl<T> is designed to only work in 3 dimensions.");
+  zero();
+}
+
+template <typename T>
+SymmetricRankFourTensorTempl<T>::SymmetricRankFourTensorTempl(const InitMethod init)
+{
+  switch (init)
+  {
+    case initNone:
+      break;
+
+    case initIdentity:
+      zero();
+      for (auto i : make_range(Ndim))
+        (*this)(i, i) = 1.0;
+      break;
+
+    case initIdentitySymmetricFour:
+      zero();
+      for (auto i : make_range(N))
+        (*this)(i, i) = 1.0;
+      break;
+
+    default:
+      mooseError("Unknown SymmetricRankFourTensorTempl<T> initialization pattern.");
+  }
+}
+
+template <typename T>
+SymmetricRankFourTensorTempl<T>::operator RankFourTensorTempl<T>()
+{
+  // Full tensor indices in the Mandel/Voigt representation
+  constexpr unsigned int g[6][6][4] = {
+      {{1, 1, 1, 1}, {1, 1, 2, 2}, {1, 1, 3, 3}, {1, 1, 2, 3}, {1, 1, 1, 3}, {1, 1, 1, 2}},
+      {{2, 2, 1, 1}, {2, 2, 2, 2}, {2, 2, 3, 3}, {2, 2, 2, 3}, {2, 2, 1, 3}, {2, 2, 1, 2}},
+      {{3, 3, 1, 1}, {3, 3, 2, 2}, {3, 3, 3, 3}, {3, 3, 2, 3}, {3, 3, 1, 3}, {3, 3, 1, 2}},
+      {{2, 3, 1, 1}, {2, 3, 2, 2}, {2, 3, 3, 3}, {2, 3, 2, 3}, {2, 3, 1, 3}, {2, 3, 1, 2}},
+      {{1, 3, 1, 1}, {1, 3, 2, 2}, {1, 3, 3, 3}, {1, 3, 2, 3}, {1, 3, 1, 3}, {1, 3, 1, 2}},
+      {{1, 2, 1, 1}, {1, 2, 2, 2}, {1, 2, 3, 3}, {1, 2, 2, 3}, {1, 2, 1, 3}, {1, 2, 1, 2}}};
```

Comment:
yeah, it should be static for sure. That's an oversight. That operator would have to take into account the mandel factors. I really don't like that idea though, as it is likely to let bad code creep in that won't benefit from the symmerty.

Reply:
> I really don't like that idea though, as it is likely to let bad code creep in that won't benefit from the symmerty.

I agree. How about adding a method with an annoying name so that people will think more before using it. Something like `getValueAndIAbsolutelyKnowWhatIAmDoing(i, j, k, l)`. Anyways, this can be a future PR if people find out that they need this.